File: src/SimpleLinearRegression.java
import java.io.BufferedReader;
import java.io.FileReader;
import java.io.IOException;
import java.util.List;
import java.util.ArrayList;
import java.util.Map;
import java.util.HashMap;

public class SimpleLinearRegression {

  Map<String, List<Integer>> data;
  String feature;
  String outcome;

  double slope;
  double intercept;

  public SimpleLinearRegression(String path, String feature, String outcome) {

      data = new HashMap<>();
      this.feature = feature;
      this.outcome = outcome;
      this.readCSV(path);
      this.buildModel();
  }

  private void readCSV(String path) {

      System.out.println("Reading csv file at " + path);
      String csvFile = path;
      String delimiter = ",";

      String line;
      boolean firstLine = true;
      List<String> featureList = new ArrayList<String>();

      try {

          BufferedReader br = new BufferedReader(new FileReader(csvFile));
          while ((line = br.readLine()) != null) {

              String[] token = line.split(delimiter);

              if (firstLine) {

                  for (int i=0; i<token.length; i++) {
                      List<Integer> values = new ArrayList<>();
                      featureList.add(token[i]);
                      data.put(token[i], values);
                  }

                  firstLine = false;
              }
              else {

                for (int i=0; i<token.length; i++) {
                    data.get(featureList.get(i)).add(Integer.parseInt(token[i]));
                }
              }
          }

          if (!data.containsKey(outcome) || !data.containsKey(feature))
              throw new Exception();

        }
        catch (Exception e) {
            System.out.println(e);
        }

    }

    private void buildModel() {
        double featureMean = 0;
        double outcomeMean = 0;

        for (int i: data.get(feature)) {
            featureMean += (double) i;
        }
        featureMean = featureMean / data.get(feature).size();

        for (int i: data.get(outcome)) {
            outcomeMean += (double) i;
        }
        outcomeMean = outcomeMean / data.get(outcome).size();

        double numerator = 0;
        double denominator = 0;

        for (int i=0; i<data.get(outcome).size(); i++) {
              numerator += ((double) data.get(outcome).get(i) - outcomeMean) *
                      ((double) data.get(feature).get(i) - featureMean);

              denominator += Math.pow(((double) data.get(feature).get(i)
                              - featureMean), 2);
        }

      slope = numerator / denominator;
      intercept = outcomeMean - slope * featureMean;

      System.out.println("Model built successfully");
    }

}

File: py/SimpleLinearRegression.py
import pandas as pd
import itertools as iter

class SimpleLinearRegression:

    def __init__(self, path, feature, outcome):
        self.path = path
        self.feature = feature
        self.outcome = outcome
        self.featureData = None
        self.outcomeData = None
        self.readCSV()
        self.buildModel()

    def readCSV(self):
        source = pd.read_csv(self.path)
        self.featureData = source[self.feature]
        self.outcomeData = source[self.outcome]

    def buildModel(self):
        featureMean = self.featureData.mean()
        outcomeMean = self.outcomeData.mean()
        num = den = 0

        for f,o in iter.izip(self.featureData, self.outcomeData):
            num += (f - featureMean) * (o - outcomeMean)
            den += (f - featureMean) * (f - featureMean)

        self.slope = num / den
        self.intercept = outcomeMean - self.slope * featureMean
        print 'Model built successfully'

    def predict(self, x):
        return self.intercept + self.slope * x

    def getSlope(self):
        return self.slope

    def getIntercept(self):
        return self.intercept

    def getTrainRMSE(self):
        return self.getRMSE(self.featureData, self.outcomeData)

    def getRMSE(self, fdata, odata):
        rss = []
        for f,o in iter.izip(fdata, odata):
            rss.append((o - self.predict(f)) ** 2)

        return sum(rss)/len(rss);

    def getTestRMSE(self, path):
        source = pd.read_csv(path)
        fdata = source[self.feature]
        odata = source[self.outcome]
        return self.getRMSE(fdata, odata)
